transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -71,16 +82,15 @@ public int get() throws Exception {
     };
     private final Callable<Integer> pendingTasksCallable = new Callable<Integer>() {
         @Override
-        public Integer call() throws Exception {
+        public Integer call() {
             return EpollEventLoop.super.pendingTasks();
         }
     };
-    private volatile int wakenUp;
-    private volatile int ioRatio = 50;
 
     EpollEventLoop(EventLoopGroup parent, Executor executor, int maxEvents,
                    SelectStrategy strategy, RejectedExecutionHandler rejectedExecutionHandler) {
         super(parent, executor, false, DEFAULT_MAX_PENDING_TASKS, rejectedExecutionHandler);
+        scheduledQueue = PlatformDependent.newMpscQueue();
```

Comment:
This is concerning.  I would have expected this to be a non threadsafe queue.   If it is threadsafe, that means that it could be concurrently modified.  That means that there is race in the code above, where calling `peek` and then `poll` could return different tasks.

Reply:
>  That means that there is race in the code above, where calling peek and then poll could return different tasks.

This is operating on a different queue. Here are the queues in play:
- `SingleThreadEventExecutor#taskQueue` - MPSC used for calls to `execute`. This is also currently used by all other transports to put process `schedule` calls on the event loop thread, but this PR changes that behavior for EPOLL.
- `AbstractScheduledEventExecutor#scheduledTaskQueue` - priority queue of scheduled `Runnable`s sorted by deadline. This is NOT thread safe and only touched on the event loop thread.
- `EpollEventLoop#scheduleQueue` - MPSC used to hold `schedule` calls outside of the event loop thread. This is drained when the event loop thread wakes up, and the `Runnable` is appended to `scheduledTaskQueue` (described above).

The goal is to decouple the scheduled task queue (`schedule` calls from outside the event loop) from the executor tasks. This allows us to drain all of the scheduled tasks to get an accurate "next expire time" before calling `epoll_wait`. In order to ensure we wakeup from outside the event loop we have `nextDeadlineNanos`, which is atomic and indicates the "next deadline", and used to know when we should call `setTimerFd` from outside (and inside) the event loop.

The class hierarchy and spread of queues is unfortunate, but since EPOLL is the only transport to decouple the `schedule` and `execute` calls I figured best to add `scheduledQueue` here for now.

Reply:
Let me add a comment describing this :)